frodo.js

```diff
@@ -1,4 +1,4 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?